src/lib/merkle_ledger/util.ml

```diff
@@ -1,38 +1,121 @@
+    Inputs.Base.t -> (Inputs.Location.t * Inputs.Account.t) list -> unit
+
+  val set_batch_accounts :
+    Inputs.Base.t -> (Inputs.Location.Addr.t * Inputs.Account.t) list -> unit
+
+  val set_all_accounts_rooted_at_exn :
+    Inputs.Base.t -> Inputs.Location.Addr.t -> Inputs.Account.t list -> unit
 end = struct
   let get_all_accounts_rooted_at_exn t address =
     let open Inputs in
-    let first_node, last_node = Addr.Range.subtree_range address in
     let result =
-      Addr.Range.fold (first_node, last_node) ~init:[] ~f:(fun bit_index acc ->
+      Location.Addr.Range.fold (Location.Addr.Range.subtree_range address)
+        ~init:[] ~f:(fun bit_index acc ->
           let account = Base.get t (location_of_addr bit_index) in
           (bit_index, account) :: acc )
     in
     List.rev_filter_map result ~f:(function
       | _, None -> None
       | addr, Some account -> Some (addr, account) )
+
+  let rec compute_affected_locations_and_hashes t locations_and_hashes acc =
+    let locations, _ = List.unzip locations_and_hashes in
+    if not @@ List.is_empty locations then
+      let height = Inputs.Location.height @@ List.hd_exn locations in
+      if height < Inputs.Depth.depth then
+        let location_to_hash_table =
+          Inputs.Location.Table.of_alist_exn locations_and_hashes
+        in
+        let _, parent_locations_and_hashes =
+          List.fold locations_and_hashes ~init:([], [])
+            ~f:(fun (processed_locations, parent_locations_and_hashes)
+               (location, hash)
+               ->
+              if
+                List.mem processed_locations location
+                  ~equal:Inputs.Location.equal
+              then (processed_locations, parent_locations_and_hashes)
+              else
+                let sibling_location = Inputs.Location.sibling location in
+                let sibling_hash =
+                  Option.value ~default:(Inputs.get_hash t sibling_location)
+                  @@ Hashtbl.find location_to_hash_table sibling_location
+                in
+                let parent_hash =
+                  let left_hash, right_hash =
+                    Inputs.Location.order_siblings location hash sibling_hash
+                  in
+                  Inputs.Hash.merge ~height left_hash right_hash
+                in
+                ( location :: sibling_location :: processed_locations
+                , (Inputs.Location.parent location, parent_hash)
+                  :: parent_locations_and_hashes ) )
+        in
+        compute_affected_locations_and_hashes t parent_locations_and_hashes
+          (List.append parent_locations_and_hashes acc)
+      else acc
+    else acc
+
+  let set_hash_batch t locations_and_hashes =
+    Inputs.set_raw_hash_batch t
+      (compute_affected_locations_and_hashes t locations_and_hashes
+         locations_and_hashes)
+
+  let set_batch t locations_and_accounts =
+    Inputs.set_raw_account_batch t locations_and_accounts ;
```

Comment:
When we do batch on a database, we should add accounts and hashes simulatenously to full atomicity. We should do this in the future.